### ISSUE TITLE
Need Google OAuth2 authentication web service end points

### ISSUE THREAD

**Issue by joshkh** [intermine/intermine#1720]
As detailed in my amazing blog post, InterMine needs a stable URL to handle Google authentication. This is to support Google users that log into BlueGenes.

See Solution 3:
https://intermineorg.wordpress.com/2017/11/09/bluegenes-oauth2-authentication-community-feedback-requested/

Requirements:

1. A stable url, such as `GET /service/auth/google?redirect_to=[some-url]`
1a. Accepts a URL parameter `redirect_to`
1b. The value of `redirect_to` is stored in the Session
1c. The url automatically redirects to the Google OAuth2 url configured for the mine.
2. A stable redirect url, such as `GET /service/auth/google-redirect`
2a. Performs the OAuth2 token exchange (existing functionality)
2b. Upon exchange, the URL fetches the user's API token from the user profile
2c. The url automatically redirects to the url found in the `redirect_to` Session variable and appends a url parameter that is the user's API token: `[some-url]?apiToken=ABC12345`

**Comment by joshkh** [comment]
Can this be made generic?

`/service/oauth2?provider=google` redirects to configured Google url
`/service/oauth2?provider=yahoo` redirects to configured Yahoo url

`/service/oauth2-redirect?provider=google` performs the workflow for exchanging the token with Google
`/service/oauth2-redirect?provider=yahoo` performs the workflow for exchanging the token with Yahoo

Steps 2b and 2c are already generic.

**Comment by vivekkrish** [comment]
Hi @joshkh -

If I'm not mistaken, there is already a stable URL which initiates the OAuth2 flow in InterMine, for example: https://apps.araport.org/thalemine/oauth2authenticator.do?provider=ARAPORT&returnto=%2Freport.do%3Fid%3D1008847

This redirects the user to a login page handled by the Araport API Manager (APIM). Upon successful authentication, the user is redirected back to ThaleMine using the configured callback: https://apps.araport.org/thalemine/oauth2callback.do?provider=ARAPORT&state=security_token%253Dacc17e50-f288-459b-a610-e0fcc0539c77%2526returnto%253D%252Freport.do%253Fid%253D1008847&code=c9803f784e7beaba655be93074c21798

All of this logic is handled in the [Callback.java standardProviderFlow()](https://github.com/intermine/intermine/blob/master/intermine/web/main/src/org/intermine/web/struts/oauth2/Callback.java#L101).

Basically, this flow would need to be forked and a new path added, which after verifying the login, grabs the users' InterMine API Token, and then redirects back to BlueGenes (rather than to an InterMine page).

**Comment by julie-sullivan** [comment]
@danielabutano this might be relevant too.

**Comment by danielabutano** [comment]
1. BG calls intermine WS /service/oauth2authenticator?provider=GOOGLE which returns a link, something like:
https://accounts.google.com/o/oauth2/v2/auth?openid.realm=http://alpha.flymine.org&scope=openid email profile&response_type=code&state={a_random_string}&client_id={the_client_id}

2. BG uses the link above to redirect to the google sign-in page
Before redirecting, BG appends the redirect_uri parameter containing the BG endpoint and the provider label (in this case GOOGLE).
The redirect URI will be an endpoint provided by BG, e.g. http://bluegenes.apps.intermine.org/alpha/oauth2callback
So the the complete link  will be:
https://accounts.google.com/o/oauth2/v2/auth?openid.realm=http://alpha.flymine.org&scope=openid email profile&response_type=code&state={a_random_string}&client_id={the_client_id}&redirect_uri=http://bluegenes.apps.intermine.org/alpha/oauth2callback?provider=GOOGLE

3. google checks if the redirect_uri received in the request is the same of the redirect URL registered on the Google APIs console, under flymine.alpha

4. After the user is signed-in in Google, google redirects to the redirect_uri
When google calls the BG endpoint the http request will be something like:
http://bluegenes.apps.intermine.org/alpha/oauth2callback?provider=GOOGLE&state={a_random_string}&code={the_authorizaion _code}&scope=email profile https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/userinfo.profile openid&authuser=0&prompt=none

5. BG calls the intermine WS /service/oauth2callback/ passing the provider, state and code extract from the request above. I do no think I need the other params.
The http request will be something like:
http://alpha.flymine.org/alpha/service/oauth2callback?provider=GOOGLE&state={a_random_string}&code={the_authorizaion _code}

**Comment by danielabutano** [comment]
added to web-properties file the additiona attribute oauth2_providers.
    `"oauth2_providers": ["GOOGLE","ELIXIR"]`
